### PR TITLE
[ccapi] Options for ccapi

### PR DESCRIPTION
**Pull request by kparichay**
Added options for ccapi

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: api/ccapi/option1.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// Layer.h (ccapi)
+class Layer {
+  public:
+    template <typename... Args>
+    Layer(Args... args) { impl = std::make_unique<LayerImpl>(args...); }
+    virtual foo() {impl->foo()};
+    virtual bar() = 0;
+
+  private:
+    std::unique_ptr<LayerImpl> impl;
```
Comment (myungjoo) [inline api/ccapi/option1.h:10]: NACK. Breaks DIP and OCP.

File: api/ccapi/option2.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    virtual bar() = 0;
+
+  private:
+    // Actual data members
+    // weights, etc
+};
+
+// layer_factory.h (ccapi)
+#include <fclayer.h>
+#include <conv2dlayer.h>
```
Comment (myungjoo) [inline api/ccapi/option2.h:15]: NACK. Breaks OCP and DIP.

File: api/ccapi/option3.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+class Layer {
+  public:
+    template <typename... Args>
+    Layer(Args... args) { std::make_unique<LayerImpl>(args...); }
+    virtual foo() {impl->foo();}
+    virtual bar() = 0;
+
+  protected:
+    // Note: different from option1 where this is private
+    std::unique_ptr<LayerImpl> impl;
```
Comment (myungjoo) [inline api/ccapi/option3.h:11]: NACK. Breaks OCP and DIP

File: api/ccapi/option3_hide_updated.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+// layer_factory.h (ccapi)
+std::unique_ptr<Layer> createLayer(string layer_type, std::vector<string> properties);
+
+// layer_factory.cpp (internal)
+#include <fclayer.h>
+#include <conv2dlayer.h>
+std::unique_ptr<Layer> createLayer(string layer_type, std::vector<string> properties) {
+  // This can be case-insensitive comparison
+  if (layer_type == "FullyConnected") {
+    FCLayer fclayer = FCLayer();
```
Comment (kparichay) [inline api/ccapi/option3_hide_updated.h:29]: constructors with `std::vector<string>` can be added but its just syntactic sugar.

**Comment by kparichay** [comment]
`api/ccapi/option3_hide_updated.h` is the style recommended by @myungjoo .

File: api/ccapi/option3_hide_updated.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    virtual foo() { /** do some work */ }
+
+  private:
+    // Actual data members
+    // weights, etc
+};
+
+// layer_factory.h (ccapi)
+std::unique_ptr<Layer> createLayer(string layer_type, std::vector<string> properties);
+
```
Comment (jijoongmoon) [inline api/ccapi/option3_hide_updated.h:22]: Do we need to vector<string> for property? Args may be?

Comment (kparichay) [reply api/ccapi/option3_hide_updated.h:22]: Args will be templated, which will need including nntrainer internal headers in API header.

Comment (zhoonit) [reply api/ccapi/option3_hide_updated.h:22]: PropertyHandler might come to a viable alternative to std::vector<string>

**Comment by kparichay** [comment]
option3_hide_updated has been finalized.
#622 implements this.

**Pull request closed without merging**